### PR TITLE
Conduit-plugin [1]: SDK

### LINKED ISSUE #35: Leveled logger for plugins
Add utility functions for plugins to create a logger with support for leveled output. Log output from plugins is already captured by Conduit and included in its own log stream, but it's currently logged without any level. The goal is to allow the plugin to decide which log level will be used for a message and possibly add structured fields.

### PR DESCRIPTION
**Pull request by lovromazgon**
### Description

This creates a new plugin SDK package that is independent of any other conduit package and is meant to be extracted into a separate repository in the future. It provides all structs a plugin implementation needs, it also adds support for structured and leveled logging in plugins.

Closes #35
Related to #65

### Quick checks:

- [X] I have followed the [Code Guidelines](https://github.com/ConduitIO/conduit/blob/main/docs/code_guidelines.md).
- [X] There is no other [pull request](https://github.com/ConduitIO/conduit/pulls) for the same update/change.
- [X] I have written unit tests.
- [X] I have made sure that the PR is of reasonable size and can be easily reviewed.

### CODE REVIEW

File: pkg/plugin/sdk/destination.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+	// function should only return an error in case a critical error happened,
+	// it is expected that all future calls to WriteAsync will fail with the
+	// same error.
+	// If the plugin caches records before writing them to the destination it
+	// needs to store the AckFunc as well and call it once the record is
+	// written.
+	// AckFunc will panic if it's called more than once.
+	// If WriteAsync returns ErrUnimplemented the SDK will fall back and call
+	// Write instead.
+	WriteAsync(context.Context, Record, AckFunc) error
```
Comment (lovromazgon) [inline pkg/plugin/sdk/destination.go:57]: I introduced `WriteAsync` and `Write` to make it easier to write connectors that don't want to cache records but instead directly write them to the destination (the comments above the functions explain the behavior of each). I'm wondering if this interface makes sense or if it's just confusing? On the other hand, using `WriteAsync` can be a bit error prone if the developer forgets to call `AckFunc` because we expect to get acks for all handled records, regardless if an error happened or not. Acceptance tests would help here, but can't completely prevent all edge cases.

Feedback on this part is especially appreciated.

File: pkg/plugin/sdk/error.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import "github.com/conduitio/conduit/pkg/foundation/cerrors"
+
+var (
+	// ErrBackoffRetry can be returned by Source.Read to signal the SDK there is
+	// no record to fetch right now and it should try again later.
+	ErrBackoffRetry = cerrors.New("backoff retry")
+
+	// ErrUnimplemented is returned in functions of plugins that don't implement
+	// a certain method.
+	ErrUnimplemented = cerrors.New("action not implemented")
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply pkg/plugin/sdk/error.go:26]: There is, but we shouldn't use it, this package is supposed to be completely independent, it will be extracted into a separate repo after all. We are only using `cerrors` here because otherwise the linter would complain, this will be easy to change once we extract the repo, and I think it's the only dependency we have towards internal Conduit packages.

File: pkg/plugin/sdk/destination.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/conduitio/conduit-plugin/cpluginv1"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+)
+
+// Destination TODO
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply pkg/plugin/sdk/destination.go:28]: No, I need to add a comment here.

File: pkg/plugin/sdk/destination.go [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+	mustEmbedUnimplementedDestination()
+}
+
+type AckFunc func(error) error
+
+func NewDestinationPlugin(impl Destination) cpluginv1.DestinationPlugin {
+	if impl == nil {
+		// prevent nil pointers
+		impl = UnimplementedDestination{}
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply pkg/plugin/sdk/destination.go:85]: Generally I agree, but I think we can get away with it here.

File: pkg/plugin/sdk/destination.go [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+	return a.ackFunc(r, stream)(err)
+}
+
+// writeAsync will call the WriteAsync function and provide the ack function to
+// the implementation without calling it.
+func (a *destinationPluginAdapter) writeAsync(ctx context.Context, r Record, stream cpluginv1.DestinationRunStream) error {
+	return a.impl.WriteAsync(ctx, r, a.ackFunc(r, stream))
+}
+
+func (a *destinationPluginAdapter) ackFunc(r Record, stream cpluginv1.DestinationRunStream) AckFunc {
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply pkg/plugin/sdk/destination.go:161]: I'll add docs.

File: pkg/plugin/sdk/record.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+	Key Data
+	// Payload holds the actual information that the record is transmitting. In
+	// a destination Payload will never be null.
+	Payload Data
+}
+
+type Position []byte
+
+type Data interface {
+	isData()
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply pkg/plugin/sdk/record.go:47]: It's so that we are sure that the only structs that implement this interface are defined in this package.

File: pkg/plugin/sdk/source.go [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+	"sync/atomic"
+	"time"
+
+	"github.com/conduitio/conduit-plugin/cpluginv1"
+	"github.com/conduitio/conduit/pkg/foundation/cerrors"
+	"github.com/jpillora/backoff"
+	"gopkg.in/tomb.v2"
+)
+
+// Source TODO
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply pkg/plugin/sdk/source.go:29]: Nope, I'll update the comment in the next commit 👍

File: pkg/plugin/sdk/serve.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+)
+
+func Serve(
+	specFactory func() Specification,
+	sourceFactory func() Source,
+	destinationFactory func() Destination,
+) {
+	err := serve(specFactory, sourceFactory, destinationFactory)
+	if err != nil {
+		_, _ = fmt.Fprintf(os.Stderr, "error running source plugin: %+v", err)
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply pkg/plugin/sdk/serve.go:32]: Nice catch 👍

File: pkg/plugin/sdk/source.go [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+	// It's guaranteed that Read will be called with a cancelled context at
+	// least once, or that the context will get cancelled while Read is running.
+	// In that case Read must stop retrieving new records from the source system
+	// and start returning records that may have already been buffered. If there
+	// are no buffered records left Read must return the context error to signal
+	// a graceful stop.
+	// After Read returns an error the function won't be called again (except if
+	// the error is ErrBackoffRetry, as mentioned above).
+	// Read can be called concurrently with Ack.
+	Read(context.Context) (Record, error)
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply pkg/plugin/sdk/source.go:55]: The consequence is that Conduit can't access a record at a random position, but we don't have a use case for that anyway. It simplifies the life of the plugin developer because they don't have to think about random access. So yeah, I also think this is a subtle but important change.

File: pkg/plugin/sdk/destination.go [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+
+func (a *destinationPluginAdapter) Configure(ctx context.Context, req cpluginv1.DestinationConfigureRequest) (cpluginv1.DestinationConfigureResponse, error) {
+	err := a.impl.Configure(ctx, req.Config)
+	return cpluginv1.DestinationConfigureResponse{}, err
+}
+
+func (a *destinationPluginAdapter) Start(ctx context.Context, req cpluginv1.DestinationStartRequest) (cpluginv1.DestinationStartResponse, error) {
+	err := a.impl.Open(ctx)
+	return cpluginv1.DestinationStartResponse{}, err
+}
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply pkg/plugin/sdk/destination.go:103]: Regarding the `Start`/`Run` discussion I'll refer you to [my comment](https://github.com/ConduitIO/conduit/pull/109#discussion_r795726653) on #109. TL;DR: the functions are split because one is a synchronous request/response while the other opens a bidirectional stream, we can't combine them (at least not in a nice way).

Regarding `Stop`/`Teardown`, there's a reason for the split here as well. Let's think about the source plugin - it is producing a stream of records and sending them to Conduit. I like to think about this stream as if it was a Go channel, essentially it is very similar - the plugin is the producer and Conduit is the consumer. And the consumer (Conduit) can't really determine when there is nothing more to read, it is the producer (plugin) who knows if there are any records left and when to actually close the stream. That's why we need `Stop`, to tell the plugin to stop producing new records and flush any records that might be cached in memory.

Now, why do we need `Teardown`? Because we have a stream of acknowledgments going back to the plugin, so even after `Stop` is called the ack stream stays open so that Conduit can deliver the acknowledgments for all remaining records. After it delivered all acknowledgments back to the plugin it still can't safely shut down the plugin because processing the acknowledgments might take some time, so that's where `Teardown` comes in - it is a signal to the plugin that Conduit won't send any more acknowledgments and wants to shut it down. The plugin can wait in that function that all acknowledgments are successfully processed and that any cleanup needed is executed, after that the plugin can safely be stopped.

Sidenote: a lot of this complexity is hidden in the SDK so that the plugin developer ideally does not have to think about it.

File: pkg/plugin/sdk/source.go [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+	<-a.readDone   // wait for read to actually stop running
+	if atomic.LoadInt32(&a.openAcks) == 0 {
+		// we aren't waiting for any further acks, let's stop the ack goroutine as well
+		a.ackCancel()
+	}
+	return cpluginv1.SourceStopResponse{}, nil
+}
+
+func (a *sourcePluginAdapter) Teardown(ctx context.Context, req cpluginv1.SourceTeardownRequest) (cpluginv1.SourceTeardownResponse, error) {
+	// TODO add a timeout and kill plugin forcefully if needed
```
[the remark this answers is not in the document]
Comment (lovromazgon) [reply pkg/plugin/sdk/source.go:230]: Good point, I'll create a follow-up for this. Actually, I need to search through all TODOs in this PR and address them, seems like I left a few.

File: pkg/plugin/sdk/source.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (hariso) [reply pkg/plugin/sdk/source.go:55]: Not having to access messages at random offsets in Kafka made it way simpler in the Kafka plugin, so much I can say.: )

File: pkg/plugin/sdk/source.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (hariso) [reply pkg/plugin/sdk/source.go:29]: Danke schön! :D

File: pkg/plugin/sdk/record.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (hariso) [reply pkg/plugin/sdk/record.go:47]: Clever "trick"!

File: pkg/plugin/sdk/error.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (hariso) [reply pkg/plugin/sdk/error.go:26]: Sounds reasonable to me, will resolve this.

**Review by hariso — APPROVED** [review]
Awesome work. 💪
